### PR TITLE
Task/multiprocess

### PR DESCRIPTION
**Pull request by maelp**
This is a first version to allow YapDatabase to work across multiple processes, you have to set the `enableMultiprocessSupport` option when creating the database. You can add a `CrossProcessNotifier` extension to be notified when another process has modified the database.

Allowing the database to work across process requires us to disable a few optimizations when the user has enabled multiprocess support. The optimizations are still available when in single process mode

@robbiehanson here is a summary of changes, would be glad if you could comment on those
## Summary of changes
- added a `busy_handler` when creating the database to make sure we can open several databases in multiprocess mode (https://github.com/yapstudios/YapDatabase/pull/283/files#diff-c9a9379c637725775619d9e0b7933ca7R75)
- added a `enableMultiprocessSupport` option to database (https://github.com/yapstudios/YapDatabase/pull/283/files#diff-2e3be8a8d3931c503fa36d74f2a407e9R273)
- we read the last snapshot from the database instead, instead of restarting from 0 (https://github.com/yapstudios/YapDatabase/pull/283/files#diff-c9a9379c637725775619d9e0b7933ca7R1350)
- replaced `busy_timeout` with a `busy_handler` which loops and sleep 50ms, outputs a log each 4 cycles if the lock was not acquired (https://github.com/yapstudios/YapDatabase/pull/283/files#diff-4fc49c6d91fb2d5322f8ee0387c2f447R72)
- updated `preReadTransaction` (https://github.com/yapstudios/YapDatabase/pull/283/files#diff-4fc49c6d91fb2d5322f8ee0387c2f447R2157)
  - added a flag `|| isMultiprocess` to force reading the latest snapshot number from the database file (instead of the last version from the singleton YapDatabase object)
  - when we get the changeset, if we miss one `snapshot` number in the list of changesets, we return `nil` instead of the array of changeset, and we clear the cache of the connection and all its extensions
  - we do this by counting the number of `relevantChangesets` which should match the increase in the snapshot number `maxSnapshot - connectionSnapshot`
  - after flushing memory, we still update the `snapshot` of the current transaction to the one that we read from the database
  - then do we also need to update the global `[database snapshot]` version? when should we update this one? there might be a case where the current connection was just updated from the database file, so it is > to the one in `[database snapshot]`
- updated `preReadWriteTransaction` (https://github.com/yapstudios/YapDatabase/pull/283/files#diff-4fc49c6d91fb2d5322f8ee0387c2f447R2442)
  - replaced `beginTransaction` with `beginImmediateTransaction` to acquire the lock if we are in multiprocess mode, so there is no concurrent readTransaction while doing a writeTransaction, which would cause issues (allowing read while there is a write transaction would require to extend the race-condition check to the multiprocess case, which would be complex)
  - in multiprocess mode, we force reading the current `snapshot` from the database file, and we update with changesets or clean cache if one is missing
- updated `postReadWriteTransaction`
  - when a `readWrite` connection makes changes and publishes a changeset, if it detected that the database was externally modified, it sends a changeset with the `YapDatabaseModifiedExternallyKey` set to `TRUE` to notify other connections and extensions that they should clean their cache
- added a `CrossProcessNotification` extension that allows to receive a `YapDatabaseModifiedExternallyNotification` when another process updates the database (it uses an identifier so you can have several databases with different identifiers if needed), but this does not transmit the changeset (https://github.com/yapstudios/YapDatabase/pull/283/files#diff-0fa70ef80dace0725d7786e5ac3cb17aR20)
- `YapDatabaseModifiedNotification`: right now we send changesets cross-process, so we don't send this modification when the database has been externally modified
- when a changeset with the `YapDatabaseModifiedExternallyKey` is sent, methods like `hasChangeForCollection`, `hasChangeForKey`, all respond `TRUE`
- when updating from a changeset (`YapDatabaseConnection.noteCommitedChangeset`), if `changeset.snapshot != connection.snapshot + 1`, we flush the cache, then update the changeset
- **QUESTION**: https://github.com/yapstudios/YapDatabase/pull/283/files#diff-4fc49c6d91fb2d5322f8ee0387c2f447R327
- **QUESTION**: what is the use of `YPDBLogTrace` and should we remove this line https://github.com/yapstudios/YapDatabase/pull/283/files#diff-cdbb67ebc7a257151869cc4413f1dc9aR146

You can see an example https://github.com/yapstudios/YapDatabase/pull/283/files#diff-4a13c0a0aa202a99f4ebe0ab5d3e6175R20

NOTE: right now the ModifiedExternally notification works by sending the `pid` as payload in libnotify, so if the current process has the same pid, it does not send the notification, and only other processes send it. Another way to go about it would be to send the snapshot number as payload, and each process check whether its `[database snapshot]` has the same value, otherwise it sends a ModifiedExternally notification (so for instance if the database instance had the time to update before receiving the libnotify message, we don't have to send another ModifiedExternally notification since the database was already updated)

### CODE REVIEW

**Comment by corristo** [comment]
Looks like tabs & spaces being mixed up.

**Comment by maelp** [comment]
@corristo thanks I will take a look

**Comment by jeremiegirault** [comment]
Please use a label "don't merge" we are still work in progress

**Comment by jeremiegirault** [comment]
@corristo can you point me where there is an issue ? will fix. thanks

File: YapDatabase/YapDatabaseOptions.m [hunk 1]
```diff
@@ -56,6 +58,7 @@ - (id)copyWithZone:(NSZone __unused *)zone
     copy->cipherPageSize = cipherPageSize;
 #endif
 	copy->aggressiveWALTruncationSize = aggressiveWALTruncationSize;
+    copy->enableMultiProcessSupport = enableMultiProcessSupport;
```
Comment (corristo) [inline YapDatabase/YapDatabaseOptions.m]: @jeremiegirault in multiple places, here's an example.

Comment (jeremiegirault) [inline YapDatabase/YapDatabaseOptions.m]: OK indeed. Xcode convention is to use spaces and Yap convention is to use tabs ?

Comment (corristo) [inline YapDatabase/YapDatabaseOptions.m]: @jeremiegirault it seems so. Let's ask @robbiehanson about project convention.

**Comment by jeremiegirault** [comment]
For cross referencing purpose, this MR intends to solve #160

**Comment by maelp** [comment]
Seems tests are not passing we have to check

**Comment by maelp** [comment]
All tests are succeeding locally, not sure why they fail on Travis

**Comment by maelp** [comment]
@robbiehanson I made the modifications we discussed, can you review the code and make comments if you think there are other part which need an update?

**Comment by maelp** [comment]
I also updated the test app to use the R-Tree extension and it seems to work correctly

**Comment by maelp** [comment]
There are a few lines with comments tagged with @robbie, if you can look at them and tell us what you think we should do in those cases

**Comment by robbiehanson** [comment]
This branch has been pulled into the "v2.9" branch, and is scheduled for inclusion in the 2.9 release.

**Comment by maelp** [comment]
:+1: Thanks man! :)

**Comment by jeremiegirault** [comment]
\o/

**Comment by jeremiegirault** [comment]
\o/

**Comment by robbiehanson** [comment]
v2.9 has officially been released, and contains this pull request.

@maelp @jeremiegirault I added a wiki page discussing [Multiprocess Support](https://github.com/yapstudios/YapDatabase/wiki/Multiprocess-Support). Feel free to add anything I missed, or improve it in any way.

**Comment by maelp** [comment]
Very nice!

**Comment by coreyfloyd** [comment]
I'm just enabling this and getting exceptions because my mappings are no longer have the correct snapshot.

This makes sense because of the statement:
> …when we get the changeset, if we miss one snapshot number in the list of changesets, we return nil instead of the array of changeset, and we clear the cache of the connection and all its extensions

So if no change sets are returned, the mappings won't update when I listen into the `YapDatabaseModifiedNotification`

I don't see mappings discussed on this thread nor do they seem to be modified in the scope of this PR. I also don't see them discussed in https://github.com/yapstudios/YapDatabase/wiki/Multiprocess-Support

Is there any guidance or an example on updating my mappings for external database modifications? Should they be updated automatically? Do I need to just recreate them?

Any help is greatly appreciated!

**Comment by coreyfloyd** [comment]
An example to illustrate…


Code executed after returning from the background after an external database modification:

```
//This will return nil by design as the cache is cleared
NSArray *notifications = [connection beginLongLivedReadTransaction];

//Since notifications is nil, this method will be a non-op
[[readConnection ext:viewName] getSectionChanges:&sectionChanges
                                                    rowChanges:&rowChanges
                                              forNotifications: notifications
                                                  withMappings:mappings];


```
At this point, the DB and the connection have moved to the new snapshot and cleared the caches but the mappings did not. 

Form this point on all I need to do to crash is write to the DB:

```

[writeConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction * _Nonnull transaction) {
//Anything at all that modifies the DB
    }];

```

Somewhere else in the app we are listening for a `YapDatabaseModifiedNotification` and updating mappings which will cause a crash:

```
//This will return new notifications from the last snapshot to the new one
NSArray *notifications = [connection beginLongLivedReadTransaction];

//And this will crash because the snapshot of the mappings is not the same as the snapshot in the first change set: 
//mappings.snapshotOfLastUpdate != [[firstChangeset objectForKey:YapDatabaseSnapshotKey] unsignedLongLongValue]
[[readConnection ext:viewName] getSectionChanges:&sectionChanges
                                                    rowChanges:&rowChanges
                                              forNotifications: notifications
                                                  withMappings:mappings];

```

The exception will read something like this:

```
preMappings.snapshotOfLastUpdate: expected(619) != found(603),
postMappings.snapshotOfLastUpdate: expected(621) != found(621), 

```

As I said in the previous comment, this makes sense to me as I don't seem to have an opportunity update the mappings when the clear the cache of the connection and all of the extensions.

**Comment by coreyfloyd** [comment]
@maelp have you seen the issue I mentioned in the issue above? Trying to figure out if I am using the API incorrectly or if there is in issue within YapDB

**Comment by afriedmanGlacier** [comment]
This is a really old issue, but I'm running into the same issue @coreyfloyd mentioned. Was there another ticket or some kind of answer for this? The docs (such as https://github.com/yapstudios/YapDatabase/wiki/Multiprocess-Support) don't seem to address this. 

Edit: Just stumbled on #504 which seems to be of help.

**Pull request merged**